### ISSUE TITLE
Hang or very slow optimization in ocamlopt with flambda

### ISSUE THREAD

**Issue by jakobbotsch** [ocaml/ocaml#10014]
We are hitting a hang or some kind of exponential blowup when compiling extracted Coq code using `ocamlopt` with flambda active. The compilation is instantaneous without flambda or when the `-Oclassic` flag is passed to `ocamlopt`.
To reproduce (sorry, this takes a while to obtain and compile all the dependencies, I'm not familiar enough with the toolchain to give a nice reduced example):
```bash
git clone https://github.com/AU-COBRA/ConCert.git
cd ConCert
git checkout cbd18da63cc1ba9ad8debe34811df8dce7457e08
opam switch create . 4.11.1+flambda
eval $(opam env)
opam repo add coq-released https://coq.inria.fr/opam/released
opam install -j 4 coq.8.11.2 coq-bignums coq-stdpp coq-quickchick
opam pin -j 4 add https://github.com/MetaCoq/metacoq.git#abc736f20020156e520e7ca4ef0557ce5f8b7db0
# Turn off the workaround
sed -i "s/ocamlopt -c -Oclassic/ocamlopt -c/g" extraction/plugin/CoqMakefile.local
make -j4
```
You should see the last `make` command getting stuck after printing
```
CAMLOPT -pack -o src/concert_extraction_plugin.cmx
```
At this point Ctrl-C does not work and `top` shows that `ocamlopt.opt` is using 100% of a core. If the workaround is then reapplied:
```bash
sed -i "s/ocamlopt -c/ocamlopt -c -Oclassic/g" extraction/plugin/CoqMakefile.local
make
```
then ocamlopt succeeds and the build finishes.

**Comment by gasche** [comment]
I am not an flambda expert, but I believe that this outcome is currently expected, not considered a bug, and unlikely to change in the short term (there are medium-term changes that could help, though).

The Flambda optimizer is designed to aggressively optimize human-written code, but it is known to blow up (or generate code that makes the backend blow up) on some generated code scenarios. It's excellent news that extracted code works well with `-Oclassic`, and I think that is a reasonable status quo for now. The point of using flambda is in large part to optimize the Coq sources themselves, that do benefit from it; I am not sure that there are actual optimization opportunities in extracted code that would benefit from higher optimization levels.

There is ongoing work on an ambitious reimplementation of the Flambda optimizer (Flambda2), and the Flambda maintainers tend to prefer to focus on that. If there are any Flambda-specific improvements to expect for extracted programs, I don't expect them to come to light before Flambda2 is moved to production.

**Comment by jakobbotsch** [comment]
> The Flambda optimizer is designed to aggressively optimize human-written code, but it is known to blow up (or generate code that makes the backend blow up) on some generated code scenarios. It's excellent news that extracted code works well with -Oclassic, and I think that is a reasonable status quo for now. The point of using flambda is in large part to optimize the Coq sources themselves, that do benefit from it; I am not sure that there are actual optimization opportunities in extracted code that would benefit from higher optimization levels.

If these are truly pathological cases that are causing exponential blowup then I would not expect it to optimize them either, but I would expect the optimizer to detect this and disable the affected optimizations or give a compiler error ("too complex code"). The workaround is fine, but having to spend several hours to debug that the cause is flambda being active and some of its optimizations blowing up on this particular code is expecting a lot from users.

**Comment by lthls** [comment]
I've managed to reproduce the issue (it takes some time...) and will look at it in depth when I have some free time. At a glance, this particular case does not look like an expected blowup, but some pack-specific problem.

**Comment by lthls** [comment]
The Ctrl-C issue appears to be a problem with `ocamlfind`, by the way. The compiler itself will handle signals normally, but for some reason when invoked through `ocamlfind` the signals from the terminal are not forwarded to the compiler.

**Comment by lthls** [comment]
I've pushed a tentative fix at #10022. The problem was that a simple rewriting of the term (that needs to happen when `pack`ing units) was exponential in the depth of function nesting. The example given has a maximum nesting depth of about 40 (after inlining), which is very uncommon; most of the time, even with generated code, the nesting depth stays low, so we didn't notice the exponential complexity of this piece of code until this issue showed up.
I think the exponential complexity is just an oversight and not a necessity though, and with #10022 the problematic compilation is done almost instantly (but given that it took more than one hour of compiling Coq modules before I could reproduce the problem, I'm not going to include this in the testsuite).

**Comment by jakobbotsch** [comment]
Thank you @lthls for the swift diagnosis and fix, that is great!
